File: geeksforgeeks/linkedlists/singlylinkedlist/N15_check_palindrome.java
import java.util.*;
class Node
{
	int data;
	Node next;
	
	Node(int d)
	{
		data = d;
		next = null;
	}
}
class N15_check_palindrome
{
	 Node head;  
	
	/* Function to print linked list */
    void printList(Node head)
    {
        Node temp = head;
        while (temp != null)
        {
           System.out.print(temp.data+" ");
           temp = temp.next;
        }  
        System.out.println();
    }
	
 
    /* Inserts a new Node at front of the list. */
    public void addToTheLast(Node node) 
	{
		if (head == null) 
		{
			head = node;
		} else 
		{
		   Node temp = head;
		   while (temp.next != null)
		   temp = temp.next;
		   temp.next = node;
		}
    }
	
	public static void main(String args[])
	{
		Scanner sc = new Scanner(System.in);
		int t = sc.nextInt();
		 
		while(t>0)
        {
			int n = sc.nextInt();
			//int k = sc.nextInt();
			Is_LinkedList_Palindrom llist = new Is_LinkedList_Palindrom();
			//int n=Integer.parseInt(br.readLine());
			int a1=sc.nextInt();
			Node head= new Node(a1);
            llist.addToTheLast(head);
            for (int i = 1; i < n; i++) 
			{
				int a = sc.nextInt(); 
				llist.addToTheLast(new Node(a));
			}
			
			GfG g = new GfG();
			if(g.isPalindrome(llist.head) == true)
			System.out.println(1);
		else
			System.out.println(0);
			t--;
		}
		
	}
}
}

/*Please note that it's Function problem i.e.
you need to write your solution in the form of Function(s) only.
Driver Code to call/invoke your function is mentioned above.*/

/* Structure of class Node is
class Node
{
	int data;
	Node next;
	
	Node(int d)
	{
		data = d;
		next = null;
	}
}*/
class GfG
{
    // if using c then we should have taken left as a double pointer
    // see .c file for this
    Node left;
    
    boolean isPalindromeUtil(Node head, Node right)    {
        left = head;
        
        if (right == null)
            return true;
            
        boolean isP = isPalindromeUtil(head, right.next);
        if (isP == false)
            return false;
            
        // check for data at this point
        if (left.data != right.data)
            return false;
            
        left = left.next;
        
        return true;
    }
    
    boolean isPalindrome(Node head) 
    {
        return isPalindromeUtil(head, head);
    }    
}

File: geeksforgeeks/linkedlists/singlylinkedlist/N15_check_palindrome.c
{
#include <stdio.h>
#include <stdlib.h>
#include<iostream>
using namespace std;
/* Link list Node */
struct Node
{
    int data;
    struct Node* next;
};
void append(struct Node** head_ref, struct Node **tail_ref, int new_data)
{
    struct Node* new_node = new Node;
    new_node->data  = new_data;
    new_node->next = NULL;
    if (*head_ref == NULL)
       *head_ref = new_node;
    else
       (*tail_ref)->next = new_node;
    *tail_ref = new_node;
}
bool isPalindrome(Node *head);
/* Driver program to test above function*/
int main()
{
  int T,i,n,l;
    cin>>T;
    while(T--){
    struct Node *head = NULL,  *tail = NULL;
        cin>>n;
        for(i=1;i<=n;i++)
        {
            cin>>l;
            append(&head, &tail, l);
        }
   	cout<<isPalindrome(head)<<endl;
    }
    return 0;
}

}


/*Please note that it's Function problem i.e.
you need to write your solution in the form of Function(s) only.
Driver Code to call/invoke your function is mentioned above.*/

/*The structure of the Node is
struct Node
{
    int data;
    struct Node* next;
};*/
/*You are required to complete this method */
bool isPalindromeUtil(Node **left, Node *right)  {
    if (right == NULL)
        return true;
        
    bool isP = isPalindromeUtil(left, right->next);
    if (isP == false)
        return false;
    
    // what happens is that now the right pointer is at the end of the 
    // ll and when we move below this point the function call stack
    // unwinds the values of the ll in the reverse order
        
    // check are the elements at the current position are equal
    if ((*left)->data != right->data)
        return false;
        
    // otherwise
    *left = (*left)->next;
    
    return true;
}

bool isPalindrome(Node *head)
{
    return isPalindromeUtil(&head, head);    
}